### PR TITLE
Implements `contains()` for `Range`s

### PR DESCRIPTION
**Pull request by desaikd**
*Description of changes:*
This PR works on changes for implementation of `contains()` in `Range`.

*Changes:*
* adds `contains()` for Range struct
* Note: TODO - implement contains for Timestamp ranges.

*Tests:*
* adds unit tests for `contains()`

### CODE REVIEW

File: src/isl/util.rs [hunk 1]
```diff
@@ -27,6 +29,224 @@ pub enum Range {
...
+                        }
+                    }
+                    (Max, _) => {
+                        unreachable!("Lower range boundary value must not be max")
+                    }
+                    (_, Min) => {
+                        unreachable!("Upper range boundary value must not be min")
+                    }
+                }
+            }
```
Comment (popematt) [inline src/isl/util.rs:242]: suggestion—The comparisons for the `(Min, Value)`, `(Value, Value)`, and `(Value, Max)` cases have repeated code. It would eliminate some of the repeated code if you refactored this so that the value we're testing is unwrapped once at the beginning of the block and so that upper and lower range bounds are considered separately (like below).
```rust
Range::Decimal(start, end) => {
    let unwrapped_value = &value.as_decimal().unwrap();
    let is_in_lower_bound = match start {
        Min => true,
        Value(start_value, boundary_type) => match start_value {
            Decimal(min_value) => match boundary_type {
                RangeBoundaryType::Inclusive => min_value <= unwrapped_value,
                RangeBoundaryType::Exclusive => min_value < unwrapped_value,
            },
            _ => unreachable!("Decimal range can only have decimals as lower and upper range boundary value"),
        },
        Max => unreachable!("Cannot have 'Max' as the lower range boundary")
    };
    // let is_in_upper_bound = match end {
    // ...
    is_in_upper_bound && is_in_lower_bound
}
```

File: src/isl/util.rs [hunk 2]
```diff
@@ -27,6 +29,224 @@ pub enum Range {
 }
 
 impl Range {
+    /// Provides a boolean value to specify whether the given value is within the range or not
+    pub fn contains(&self, value: &OwnedElement) -> IonSchemaResult<bool> {
```
Comment (popematt) [inline src/isl/util.rs:33]: issue—from the spec (emphasis added):
> A number range includes all values of any numeric type (i.e. `float`, `int`, and `decimal`) that fall within that range mathematically. When testing a value for inclusion, the range bounds and the value that is being tested _will all be converted to `decimal` values for the sake of comparison_. `number ranges` do not include values of any other type.

...but this implementation does not convert any of the other number types to a decimal.

I think this is also a good opportunity to separate out the normalization to decimal from the actual comparison, which will help eliminate a lot of very similar code that repeats the comparison logic for each type.

----

question—given that numbers are treated identically when it comes to comparison, do you think it would be worth refactoring `Range` into `NumberRange` and `TimestampRange`?

File: src/isl/mod.rs [hunk 3]
```diff
@@ -337,4 +338,135 @@ mod isl_tests {
...
+            vec![5.into(), 0.into(), 10.into()],
+            vec![(-5).into(), 11.into()]
+        ),
+        case::int_range_with_min(
+            load_range(
+            r#"
+                range::[min, 10]
+            "#
+            ),
+            vec![5.into(), (-5).into(), 0.into()],
```
Comment (popematt) [inline src/isl/mod.rs:361]: suggestion—it would be good to also add valid test cases for `0e0`, `0d0`, `-0e0`, `-0d0`, `5d0`, `5e0`, `10e0`, and `10d0`.

File: src/isl/mod.rs [hunk 4]
```diff
@@ -337,4 +338,135 @@ mod isl_tests {
...
+        case::float_range(
+            load_range(
+            r#"
+                range::[1e2, 5e2]
+            "#
+            ),
+            vec![2e2.into(), 1e2.into(), 5e2.into()],
+            vec![(-1e2).into() ,1e1.into(), 6e2.into()]
+        ),
+        case::float_range_with_min(
```
Comment (popematt) [inline src/isl/mod.rs:427]: suggestion—add test case for lowest possible float value

File: src/isl/mod.rs [hunk 5]
```diff
@@ -337,4 +338,135 @@ mod isl_tests {
...
+            vec![2e5.into(), (-2e5).into()],
+            vec![3e5.into()]
+        ),
+        case::float_range_with_max(
+            load_range(
+            r#"
+                range::[1e5, max]
+            "#
+            ),
+            vec![1e5.into(), 5e5.into(), 1e6.into()],
```
Comment (popematt) [inline src/isl/mod.rs:442]: suggestion—add valid test case for max possible float value

File: src/isl/mod.rs [hunk 6]
```diff
@@ -337,4 +338,135 @@ mod isl_tests {
...
+            vec![(-5e5).into(), 1e2.into()]
```
Comment (popematt) [inline src/isl/mod.rs:443]: suggestion—add invalid test case for `nan`, `+inf`, and `-inf`

**Review by popematt — CHANGES_REQUESTED** [review]
As I was reviewing this and suggesting more test cases, it occurred to me that these test cases should be covered by the `ion-schema-tests`. Is it possible to integrate with `ion-schema-tests` even if we can't run all of them right now? (Maybe the test runner can `unwind` any panics that are caused by a `todo!` and mark the test as "skipped" somehow.)

There might be a lot of effort saved if you could develop directly against `ion-schema-tests` instead of having to replicate _every_ possible edge case with unit tests.

File: src/isl/util.rs [hunk 7]
```diff
@@ -27,6 +29,224 @@ pub enum Range {
...
+        use RangeBoundaryValue::*;
+        use RangeBoundaryValueType::*;
+        match self {
+            Range::Integer(start, end) => {
+                match (start, end) {
+                    (Min, Max) => {
+                        unreachable!("Range boundaries can not be min and max together (i.e. range::[min, max] is not allowed)")
+                    },
```
[the remark this answers is not in the document]
Comment (desaikd) [reply src/isl/util.rs:41]: Yes. Many of these `unreachable!` statements are because they have been already checked while constructing the `Range`.

File: src/isl/util.rs (hunk 2 above)
Comment (desaikd) [reply src/isl/util.rs:33]: I can certainly normalize ranges to decimal for comparison. 

From my understanding, you are suggesting to refactor `Range` into `NumberRange` and `TimestampRange` as all number `Range`s will be converted to `Decimal` under the hood. But from the reference of `ion-schema-kotlin` [implementation](https://github.com/amzn/ion-schema-kotlin/blob/master/src/com/amazon/ionschema/internal/util/RangeIonTimestamp.kt#L53) and from above statement it does seem like all types of ranges are being converted to `Decimal`, so maybe a single method implementation could be used for all of them. Unless, there was a specific reason you suggested this?

**Comment by desaikd** [comment]
> As I was reviewing this and suggesting more test cases, it occurred to me that these test cases should be covered by the `ion-schema-tests`. Is it possible to integrate with `ion-schema-tests` even if we can't run all of them right now? (Maybe the test runner can `unwind` any panics that are caused by a `todo!` and mark the test as "skipped" somehow.)
> 
> There might be a lot of effort saved if you could develop directly against `ion-schema-tests` instead of having to replicate _every_ possible edge case with unit tests.

Yes, I can integrate with `ion-schema-tests` in a separate PR. As for `Range#contains` we would need separate unit test cases from `ion-schema-tests` as it doesn't include `Range#contains` specific tests.

**Comment by popematt** [comment]
> > As I was reviewing this and suggesting more test cases, it occurred to me that these test cases should be covered by the `ion-schema-tests`. Is it possible to integrate with `ion-schema-tests` even if we can't run all of them right now? (Maybe the test runner can `unwind` any panics that are caused by a `todo!` and mark the test as "skipped" somehow.)
> > There might be a lot of effort saved if you could develop directly against `ion-schema-tests` instead of having to replicate _every_ possible edge case with unit tests.
> 
> Yes, I can integrate with `ion-schema-tests` in a separate PR for all ISL constraints. As for `Range#contains` we would need separate unit test cases from `ion-schema-tests` as it doesn't include `Range#contains` specific tests.

You're right that it wouldn't contain tests specifically for `Range#contains`, but it does contain test cases that could have caught the some things here (via the `valid_values` tests). I'm certainly not suggesting that we do away with unit testing. Anyway, this is not something that needs to be resolved for this PR—it was just an observation that I had. Sorry to distract from the main point of the PR.

File: src/isl/util.rs (hunk 2 above)
Comment (popematt) [reply src/isl/util.rs:33]: > From my understanding, you are suggesting to refactor Range into NumberRange and TimestampRange as all number Ranges will be converted to Decimal under the hood. But from the reference of ion-schema-kotlin implementation and from above statement it does seem like all types of ranges are being converted to Decimal, so maybe a single method implementation could be used for all of them. Unless, there was a specific reason you suggested this?

I don't have a clearly formed idea about this—I was just throwing an idea out there to see if you had an opinion on it. Don't worry about it unless it seems to you like it would make the implementation simpler.

Comment (desaikd) [reply src/isl/util.rs:33]: Sure. Thanks. I will try to restructure this for using `Decimal` comparison and see if we can change something here.

File: src/isl/util.rs [hunk 8]
```diff
@@ -27,6 +29,224 @@ pub enum Range {
...
 impl Range {
+    /// Provides a boolean value to specify whether the given value is within the range or not
+    pub fn contains(&self, value: &OwnedElement) -> IonSchemaResult<bool> {
+        use RangeBoundaryValue::*;
+        use RangeBoundaryValueType::*;
+        match self {
+            Range::Integer(start, end) => {
+                match (start, end) {
+                    (Min, Max) => {
+                        unreachable!("Range boundaries can not be min and max together (i.e. range::[min, max] is not allowed)")
```
Comment (zslayton) [inline src/isl/util.rs:40]: Are all of these `unreachable!` states being checked when the `Range` is being constructed? For example, is this `unreachable!` because we proved that the range wasn't `range::[min, max]` when we parsed it?

I'm confirming that this is a code path we genuinely don't expect to encounter. If this is the first time we're checking for this case, we should return an error instead.

Comment (desaikd) [reply src/isl/util.rs:40]: Yes, its been checked while the construction of the range.

File: src/isl/util.rs [hunk 9]
```diff
@@ -35,210 +35,147 @@ impl Range {
...
-                                        Ok(int_start_value < &value.as_any_int().unwrap().as_i64().unwrap() || (*boundary_type == RangeBoundaryType::Inclusive && int_start_value == &value.as_any_int().unwrap().as_i64().unwrap()))
-                                    }
-                                    AnyInt::BigInt(big_int_start_value) => {
-                                        Ok(big_int_start_value < &value.as_any_int().unwrap().as_big_int().unwrap() || (*boundary_type == RangeBoundaryType::Inclusive && big_int_start_value == value.as_any_int().unwrap().as_big_int().unwrap()))
-                                    }
-                                }
-                            },
-                            _ => unreachable!("Integer range can only have integers as lower and upper range boundary value")
-                        }
+                let unwrapped_value = value.as_any_int().unwrap();
```
Comment (zslayton) [inline src/isl/util.rs:38]: A few thoughts here:

1. If this is a `Range<integer>` and I ask whether a `float` is in the range, I think this should return an `Err` instead of panicking.
2. I think eventually we'll need an enum variant for `Range<number>` [as described here](https://amzn.github.io/ion-schema/docs/spec.html#rangenumber). This is where we'd convert any/all types to `decimal` during comparison. We can do this in a follow-on PR.
3. (Just an observation, no change needed.) We don't have any use for `value` after it gets unwrapped. This seems like a situation where variable shadowing would work nicely.

```rust
let value = value.as_any_int().unwrap();
```

This creates a new variable with the same name (`value`) but a different type (`AnyInt` instead of `Option<AnyInt>`).

Comment (desaikd) [reply src/isl/util.rs:38]: For 2, I meant to create an issue to track it. Here's the reference issue: https://github.com/amzn/ion-schema-rust/issues/58

File: src/isl/util.rs [hunk 10]
```diff
@@ -35,12 +35,16 @@ impl Range {
         use RangeBoundaryValueType::*;
         match self {
             Range::Integer(start, end) => {
-                let unwrapped_value = value.as_any_int().unwrap();
+                let value = value.as_any_int().ok_or_else(|| {
+                    return invalid_schema_error_raw(
+                        "Integer ranges can only have integer value for validation",
+                    );
+                })?;
```
Comment (zslayton) [inline src/isl/util.rs:42]: Here and elsewhere in the PR, you can simplify this to:
```rust
let value = value.as_any_int().ok_or_else(|| invalid_schema_error_raw("..."))?;
```
(That is: you can remove the explicit `return` and braces.)

**Review by zslayton — APPROVED** [review]

**Review by popematt — APPROVED** [review]
